File: src/main/java/com/dh/ondot/notification/domain/service/EmergencyAlertService.java
package com.dh.ondot.notification.domain.service;

import com.dh.ondot.core.util.TimeUtils;
import com.dh.ondot.notification.domain.EmergencyAlert;
import com.dh.ondot.notification.domain.dto.EmergencyAlertDto;
import com.dh.ondot.notification.domain.repository.EmergencyAlertRepository;
import com.dh.ondot.notification.infra.emergency.EmergencyAlertApi;
import com.dh.ondot.notification.infra.emergency.EmergencyAlertDtoMapper;
import com.dh.ondot.notification.infra.emergency.EmergencyAlertJsonExtractor;
import com.fasterxml.jackson.databind.JsonNode;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.*;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class EmergencyAlertService {
    private final EmergencyAlertApi emergencyAlertApi;
    private final EmergencyAlertJsonExtractor emergencyAlertJsonExtractor;
    private final EmergencyAlertDtoMapper emergencyAlertDtoMapper;
    private final EmergencyAlertRepository emergencyAlertRepository;

    public List<EmergencyAlertDto> fetchAlertsByDate(LocalDate date) {
        String rawJson = emergencyAlertApi.fetchAlertsByDate(date);
        JsonNode alertsNode = emergencyAlertJsonExtractor.extractAlerts(rawJson);
        return emergencyAlertDtoMapper.toDto(alertsNode);
    }

    @Transactional(readOnly = true)
    public String getIssuesByAddress(String roadAddress) {
        String regionKey = extractRegionKey(roadAddress);
        String provinceKey = extractProvince(roadAddress);
        String allRegion = provinceKey + " 전체";

        LocalDate today = TimeUtils.nowSeoulDate();
        Instant from = TimeUtils.toInstant(today.atStartOfDay());
        Instant to = TimeUtils.toInstant(today.plusDays(1).atStartOfDay());
        List<EmergencyAlert> alerts = emergencyAlertRepository.findAllByCreatedAtBetween(from, to);

        List<String> contents = alerts.stream()
                .filter(a -> matchesRegionCSV(a.getRegionName(), regionKey, allRegion))
                .map(EmergencyAlert::getContent)
                .collect(Collectors.toList());

        return String.join("\n", contents);
    }

    // “서울특별시 동작구 흑석로23-2” → “서울특별시 동작구”
    private String extractRegionKey(String roadAddress) {
        String[] tok = roadAddress.split("\\s+");
        if (tok.length >= 2) return tok[0] + " " + tok[1];
        if (tok.length == 1) return tok[0];
        return "";
    }

    // “서울특별시 동작구 흑석로23-2” → “서울특별시”
    private String extractProvince(String roadAddress) {
        String[] tok = roadAddress.split("\\s+");
        return tok.length >= 1 ? tok[0] : "";
    }

    private boolean matchesRegionCSV(String regionCsv, String regionKey, String allRegion) {
        for (String part : regionCsv.split(",")) {
            String r = part.trim();
            if (r.equals(regionKey) || r.equals(allRegion)) {
                return true;
            }
        }
        return false;
    }
}

File: src/main/kotlin/com/dh/ondot/notification/domain/service/EmergencyAlertService.kt
package com.dh.ondot.notification.domain.service

import com.dh.ondot.core.util.TimeUtils
import com.dh.ondot.notification.domain.dto.EmergencyAlertDto
import com.dh.ondot.notification.domain.repository.EmergencyAlertRepository
import com.dh.ondot.notification.infra.emergency.EmergencyAlertApi
import com.dh.ondot.notification.infra.emergency.EmergencyAlertDtoMapper
import com.dh.ondot.notification.infra.emergency.EmergencyAlertJsonExtractor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate

@Service
class EmergencyAlertService(
    private val emergencyAlertApi: EmergencyAlertApi,
    private val emergencyAlertJsonExtractor: EmergencyAlertJsonExtractor,
    private val emergencyAlertDtoMapper: EmergencyAlertDtoMapper,
    private val emergencyAlertRepository: EmergencyAlertRepository,
) {

    fun fetchAlertsByDate(date: LocalDate): List<EmergencyAlertDto> {
        val rawJson = emergencyAlertApi.fetchAlertsByDate(date)
        val alertsNode = emergencyAlertJsonExtractor.extractAlerts(rawJson)
        return emergencyAlertDtoMapper.toDto(alertsNode)
    }

    @Transactional(readOnly = true)
    fun getIssuesByAddress(roadAddress: String): String {
        val regionKey = extractRegionKey(roadAddress)
        val provinceKey = extractProvince(roadAddress)
        val allRegion = "$provinceKey 전체"

        val today = TimeUtils.nowSeoulDate()
        val from = TimeUtils.toInstant(today.atStartOfDay())
        val to = TimeUtils.toInstant(today.plusDays(1).atStartOfDay())
        val alerts = emergencyAlertRepository.findAllByCreatedAtBetween(from, to)

        val contents = alerts
            .filter { a -> matchesRegionCSV(a.regionName, regionKey, allRegion) }
            .map { it.content }

        return contents.joinToString("\n")
    }

    // "서울특별시 동작구 흑석로23-2" → "서울특별시 동작구"
    private fun extractRegionKey(roadAddress: String): String {
        val tok = roadAddress.split("\\s+".toRegex())
        return when {
            tok.size >= 2 -> "${tok[0]} ${tok[1]}"
            tok.size == 1 -> tok[0]
            else -> ""
        }
    }

    // "서울특별시 동작구 흑석로23-2" → "서울특별시"
    private fun extractProvince(roadAddress: String): String {
        val tok = roadAddress.split("\\s+".toRegex())
        return if (tok.isNotEmpty()) tok[0] else ""
    }

    private fun matchesRegionCSV(regionCsv: String, regionKey: String, allRegion: String): Boolean {
        for (part in regionCsv.split(",")) {
            val r = part.trim()
            if (r == regionKey || r == allRegion) {
                return true
            }
        }
        return false
    }
}
